rfcs/0019-storing-client-data.md

```diff
@@ -0,0 +1,1033 @@
+    previous_hash: string
+  }
+
+  val compute_receipt : ~prev:t -> User_command.t -> t
+end
+
+module Receipt_chain_database = struct
+  type t = (Receipt_chain_hash.t, Receipt_chain_hash.t)Key_value_store.t
+
+  val add : previous:Receipt_chain_hash.t -> t -> User_command.t -> unit
+
+  val prove : t -> proving_receipt:Receipt_chain_hash.t -> resulting_receipt:Receipt_chain_hash.t ->
+(Receipt_chain_hash.t, User_command.t) Payment_proof.t Base.Or_error.t
+
+  val verify : t -> resulting_receipt:Receipt_chain_hash.t ->
+(Receipt_chain_hash.t, User_command.t) Payment_proof.t ->
+(unit, Error.t) Pervasives.result
+end
+```
+
+These containers should also have performant pagination queries, which is essentially a single contiguous range of sorted elements. Namely, we order blocks and transactions by some notion of time. These metrics of time vary based on a client's needs.
+
+Blocks and transactions can be ordered by the time it is first seen. The first time we see a block is considered to be the time it gets added to the `transition_frontier`. The first time we see a transaction would be the first time we see it in the `transaction_pool` or when it gets added into a block in the `transition_frontier`.
+
+We can also order these objects by the lexicographic ordering of block length, epoch and slot. We will denote this ordering as `block_compare`. `block_compare` length can arguably be a better ordering for blocks because clients can receive blocks at different times, but what remains constant is the `block_compare` that the block appears in the blockchain.
+
+The ocaml code to sort blocks can be seen as the following:
+
+```ocaml
+val block_compare = lexicographic [
+  Comparable.lift Length.compare ~f:Block.block_length;
+  Comparable.lift Epoch.compare ~f:Block.epoch;
+  Comparable.left Epoch.Slot.compare ~f:Block.epoch_slot
+]
+```
+
+Additionally, we can sort transactions by the minimum `block_compare` of the blocks that they appear in. Typically, a client would expect to receive a transaction at the chain that honest peers are building. This honest chain should have a small `block_compare` to other blocks. If the transactions do not appear in an block, they will not appear in the sorting.
+
+A drawback with sorting blocks and transaction based on the time that they are first seen is that if the client wants to store an arbitrary transaction and block, then they would not have the extra burden of tagging these objects with a timestamp for pagination. `block_compare` alleviates this issue.
+
+## SQL Database
+
+The relational databases that we should use should be agnostic. There are different use cases for using a light database, such as SQLite, and heavier database, such as Postgres. Postgres should be used when we have an archive node or a block explorer trying to make efficient writes to store everything that it hears from the network. It also supports built-in subscriptions that would be convient for clients to listen to.
```

Comment:
Suggestion:

This architecture supports many SQL databases and even excluding a database entirely. When it is included, there are different use cases for using a light database, such as SQLite, and heavier database, such as Postgres. Postgres should be used when we have an archive node or a block explorer trying to make efficient writes to store everything that it hears from the network. It also supports built-in subscriptions that would be convinient for clients to listen to.